source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,2664 @@
+allow test runners to determine whether a particular test file is supported.
+
+New tests files SHOULD always be written using the latest major version of this
+specification; however, test files SHOULD be conservative in the minor version
+they specify (as noted in `schemaVersion`_).
+
+
+JSON Schema Validation
+~~~~~~~~~~~~~~~~~~~~~~
+
+Each major version of this specification will have a corresponding JSON schema
+for its most recent minor version (e.g. ``schema-1.1.json``). A JSON schema for
+a particular minor version MUST be capable of validating any and all test files
+in that major version series up to and including the minor version. For example,
+``schema-2.1.json`` should validate test files with `schemaVersion`_ "2.0" and
+"2.1", but would not be expected to validate "1.0", "2.2", or "3.0".
+
+The JSON schema MUST remain consistent with the `Test Format`_ section. If and
+when a new major version is introduced, the `Breaking Changes`_ section MUST be
+updated and any JSON schema(s) for a previous major version(s) MUST remain
+available so that older test files can still be validated.
+
+`Ajv <https://ajv.js.org/>`__ MAY be used to programmatically validate both YAML
+and JSON files using the JSON schema. The JSON schema MUST NOT use syntax that
+is unsupported by this tool, which bears mentioning because there are multiple
+versions of the
+`JSON schema specification <https://json-schema.org/specification.html>`__.
+
+
+Test Runner Support
+~~~~~~~~~~~~~~~~~~~
+
+Each test file defines a `schemaVersion`_, which test runners will use to
+determine compatibility (i.e. whether and how the test file will be
+interpreted). Test files are considered compatible with a test runner if their
+`schemaVersion`_ is less than or equal to a supported version in the test
+runner, given the same major version component. For example:
+
+- A test runner supporting version 1.5.1 could execute test files with versions
+  1.0 and 1.5 but *not* 1.6 and 2.0.
+- A test runner supporting version 2.1 could execute test files with versions
+  2.0 and 2.1 but *not* 1.0 and 1.5.
+- A test runner supporting *both* versions 1.5.1 and 2.0 could execute test
+  files with versions 1.4, 1.5, and 2.0, but *not* 1.6, 2.1, or 3.0.
+- A test runner supporting version 2.0.1 could execute test files with versions
+  2.0 and 2.0.1 but *not* 2.0.2 or 2.1. This example is provided for
+  completeness, but test files SHOULD NOT need to refer to patch versions (as
+  previously mentioned).
+
+Test runners MUST NOT process incompatible files but MAY determine how to handle
+such files (e.g. skip and log a notice, fail and raise an error). Test runners
+MAY support multiple schema versions (as demonstrated in the example above).
+
+
+Impact of Spec Changes on Schema Version
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Backwards-breaking changes SHALL warrant a new major version. These changes
+include, but are not limited to:
```

Comment:
I would like to consider whether new operations (e.g. for a new helper method, like aggregate) or a new option to an existing operation (e.g. adding hint support to aggregate), should bump the minor version as well.

Changes like these, similar to the ones listed below, don’t invalidate existing tests, but do require a change to the test runner in order to correctly execute the new tests.  So it seems like a distinction without a difference to require a minor version bump for one and not the other.

Jeremy mentioned to me that going this route would likely entail defining all operations/arguments in the json schema and likely adding a type to each operation (e.g. `type: collection`) since the schema can’t very well infer context from `object` which is an entity name.

Another option is to create sub-versions, e.g. `version: {schema: 1.0, crud: 1.1}`.

Reply:
I'm not sure how to effectively handle this short of enforcing the entire API in the JSON schema. In that case, we'd have to go beyond operations/arguments and also cover things like URI options for configuring client entities.

> Another option is to create sub-versions, e.g. `version: {schema: 1.0, crud: 1.1}.`

This seems like it'd add quite a bit more complexity -- if not to the test runner then at least our own processes for tracking versions. We don't have a consistent process for tracking driver spec compliance or even versioning spec documents. Tracking driver support for server versions may be more feasible (e.g. if we expressed something like "requires 4.4-compat driver" in `runOnRequirements`), but even that has its shortcomings and edge cases where a driver might be late to implement a particular feature or cannot due to some technical reason.

Reply:
I'd also be interested in finding some way to cause new operations to require a version bump. It seems a little strange that a new test could require changes to the test runner but not cause any version bump at all.

Reply:
Perhaps it would be helpful, if only for humans reading the spec, to distinguish two sets of versions:

- The test format version
- The set of allowed operators and the definition of each operator

I agree with @jyemin in that in practice, adding operators requires test runner changes, therefore I support new operators increasing *some* version number.

On the topic of version numbers, I know that all of our specs have them, and personally I don't recall ever using a spec's version number for anything. I have always looked at the diff for the particular change I am implementing and in general I sync test files completely. So perhaps version numbers are not really that important.

I am less familiar with json schema but, assuming it is not practical to validate the operators, I think it is worthwhile to consider whether not validating operators makes us worse off than the present test format and runner situation (where nothing is validated at all).

What about, as a possible compromise, the following approach:

- Any change that requires a change in the tests' json schema is a major version bump (this includes making a required field optional).
- Adding new operators while not changing json schema is a minor version bump.

This way the version number is useful for engineers changing the spec format and it means something for engineers adding operations. On the implementation side the version number indicates whether no changes to test runners are required, whether changes are limited to adding operators or whether there are other changes.

The patch number in this system can perhaps be done away with entirely.

This approach would result in more major versions but I think this is not a significant problem in practice given that addition of an operator does practically require changing the test runner, so I think the test runners ultimately would end up supporting either a range of versions (including major versions) anyway or they would support a single minor version.

Reply:
Something that occurred to me while talking this over with @alcaeus. In https://github.com/mongodb/specifications/pull/846#discussion_r475870493, @jyemin was suggesting that a test runner should raise an error when fed a test file that is not supported due to its `schemaVersion`. I currently allow test runners to handle this as they like, but was open to providing SHOULD guidance to encourage failing for more visibility.

In https://github.com/mongodb/specifications/pull/846#discussion_r475877714, @jyemin asked me to clarify behavior for unsupported keys in `operation.arguments` and I updated the spec to require test runners to raise an error (as was already done for unsupported operations). In that case, tests will fail if they encounter an unsupported operation or argument.

Unless I'm missing something, there seems to be no practical difference between whether we extend the effort (now and in the future) to track all possible operations/arguments in the JSON schema and tie it to the `schemaVersion`.

Reply:
> There seems to be no practical difference between whether we extend the effort (now and in the future) to track all possible operations/arguments in the JSON schema and tie it to the schemaVersion.

By "no practical difference", do you mean that the end result is an error / test failure either way, all that changes is where the error is reported (when comparing schema version vs. when encountering an unsupported operation or argument)?

While that is true, it does then require that we actually download and run test files to determine if we can support them, right? If the change were communicated by a schema version, we could just compare the version with the known supported schema version(s) for our runner.

Reply:
Following on @kmahar's comment, which I agree with, this would seem to have ramification for automated spec testing, as it could be useful to determine ahead of time, or at least clearly report, when a driver can't run a given test due to an obvious known incompatibility. 

@mbroadst perhaps you should weigh in on whether this would be helpful.